src/pocketmine/command/CommandReader.php

```diff
@@ -19,58 +19,138 @@
  *
 */
 
+declare(strict_types = 1);
+
 namespace pocketmine\command;
 
 use pocketmine\Thread;
 
 class CommandReader extends Thread{
-	private $readline;
+
+	const TYPE_READLINE = 0;
+	const TYPE_STREAM = 1;
+	const TYPE_PIPED = 2;
+
 	/** @var \Threaded */
 	protected $buffer;
 	private $shutdown = false;
-	private $streamBlocking = false;
+	private $type = self::TYPE_STREAM;
+	private $lastLine = -1;
 
 	public function __construct(){
 		$this->buffer = new \Threaded;
 		$opts = getopt("", ["disable-readline"]);
-		$this->readline = (extension_loaded("readline") and !isset($opts["disable-readline"]));
+
+		if((extension_loaded("readline") and !isset($opts["disable-readline"]) and !$this->isPipe(STDIN))){
+			$this->type = self::TYPE_READLINE;
```

Comment:
it would be nice to output a debug message about what type is being used. that would help in many cases.